pkg/controller/volume/persistentvolume/binder_test.go

```diff
@@ -517,6 +519,141 @@ func TestSync(t *testing.T) {
+			withVolumeVolumeMode(&modeBlock, newVolumeArray("volume14-6-1", "10Gi", "", "", v1.VolumePending, v1.PersistentVolumeReclaimRetain, classEmpty)),
+			withClaimVolumeMode(&modeFile, newClaimArray("claim14-6-1", "uid14-6-1", "10Gi", "", v1.ClaimPending, nil)),
+			withClaimVolumeMode(&modeFile, newClaimArray("claim14-6-1", "uid14-6-1", "10Gi", "", v1.ClaimPending, nil)),
+			[]string{"Normal FailedBinding"},
+			noerrors, testSyncClaim,
+		},
+		{
+			// failed syncVolume do not bind when volumeModes do not match
+			// for example claim requests block but volume is unspecified (filesystem by default)
+			"14-6-2 - do not bind when volumeModes do not match, unspecified volume and requested block for claim",
+			withVolumeVolumeMode(nil, newVolumeArray("volume14-6-2", "10Gi", "", "", v1.VolumePending, v1.PersistentVolumeReclaimRetain, classEmpty)),
+			withVolumeVolumeMode(nil, newVolumeArray("volume14-6-2", "10Gi", "", "", v1.VolumePending, v1.PersistentVolumeReclaimRetain, classEmpty)),
+			withClaimVolumeMode(&modeBlock, newClaimArray("claim14-6-2", "uid14-6-2", "10Gi", "", v1.ClaimPending, nil)),
+			withClaimVolumeMode(&modeBlock, newClaimArray("claim14-6-2", "uid14-6-2", "10Gi", "", v1.ClaimPending, nil)),
+			[]string{"Normal FailedBinding"},
+			noerrors, testSyncClaim,
+		},
+		{
+			// syncVolume bind when pv is prebound and volumeModes match
+			"14-7 - bind when pv volume is prebound and volumeModes match",
+			withVolumeVolumeMode(&modeBlock, newVolumeArray("volume14-7", "10Gi", "", "claim14-7", v1.VolumePending, v1.PersistentVolumeReclaimRetain, classEmpty)),
+			withVolumeVolumeMode(&modeBlock, newVolumeArray("volume14-7", "10Gi", "uid14-7", "claim14-7", v1.VolumeBound, v1.PersistentVolumeReclaimRetain, classEmpty)),
+			withClaimVolumeMode(&modeBlock, newClaimArray("claim14-7", "uid14-7", "10Gi", "", v1.ClaimPending, nil)),
+			withClaimVolumeMode(&modeBlock, newClaimArray("claim14-7", "uid14-7", "10Gi", "volume14-7", v1.ClaimBound, nil, annBindCompleted)),
+			noevents, noerrors, testSyncClaim,
+		},
+		{
+			// failed syncVolume do not bind when pvc is prebound to pv with mismatching volumeModes
+			"14-8 - do not bind when pvc is prebound to pv with mismatching volumeModes",
+			withVolumeVolumeMode(&modeBlock, newVolumeArray("volume14-8", "10Gi", "", "", v1.VolumePending, v1.PersistentVolumeReclaimRetain, classEmpty)),
+			withVolumeVolumeMode(&modeBlock, newVolumeArray("volume14-8", "10Gi", "", "", v1.VolumePending, v1.PersistentVolumeReclaimRetain, classEmpty)),
+			withClaimVolumeMode(&modeFile, newClaimArray("claim14-8", "uid14-8", "10Gi", "volume14-8", v1.ClaimPending, nil)),
+			withClaimVolumeMode(&modeFile, newClaimArray("claim14-8", "uid14-8", "10Gi", "volume14-8", v1.ClaimPending, nil)),
+			[]string{"Normal FailedBinding"},
+			noerrors, testSyncClaim,
+		},
+		{
+			// syncVolume binds when pvc is prebound to pv with matching volumeModes block
```

Comment:
also pvc is prebound but mismatching volumeModes

Reply:
also prebound, matching volumeModes, but mismatched storageclass

Reply:
doesn't 14-8 cover pvc is prebound but mismatching volumeModes?

Adding 14-5-2 for prebound pvc with matching modes but mismatched class

Reply:
sorry, i mean prebound pv with mismatching volume modes

Reply:
ah, adding 14-8-1 test